gdal/frmts/netcdf/netcdfsgwriterutil.cpp

```diff
@@ -277,547 +408,554 @@ namespace nccfdriver
-
-    SGWriter_Exception_NCDefFailure::SGWriter_Exception_NCDefFailure(const char * layer_name, const char * failure_name, const char * failure_type)
-    : msg(sgwe_msg_builder(layer_name, failure_name, failure_type,
-            "could not be defined in the dataset (definition failure)."))
-    {}
+    std::shared_ptr<OGR_SGFS_Transaction> WTransactionLog::pop()
+    {
+         if(log == nullptr)
+             return std::shared_ptr<OGR_SGFS_Transaction>(nullptr);
+
+         int varId;
+         nc_type ntype;
+         size_t itemsread;
+         itemsread = VSIFReadL(&varId, sizeof(int), 1, log);
+         itemsread &= VSIFReadL(&ntype, sizeof(nc_type), 1, log);
+
+         // If one of the two reads failed, then return nullptr
+         if(!itemsread)
+             return std::shared_ptr<OGR_SGFS_Transaction>(nullptr);
+
+         // If not, continue on and parse additional fields
+         switch(ntype)
+         {
+             // NC-3 Primitives
+             case NC_BYTE:
+                 return genericLogDataRead<OGR_SGFS_NC_Byte_Transaction, signed char>(varId, log);
+             case NC_SHORT:
+                 return genericLogDataRead<OGR_SGFS_NC_Short_Transaction, short>(varId, log);
+             case NC_INT:
+                 return genericLogDataRead<OGR_SGFS_NC_Int_Transaction, int>(varId, log);
+             case NC_FLOAT:
+                 return genericLogDataRead<OGR_SGFS_NC_Float_Transaction, float>(varId, log);
+             case NC_DOUBLE:
+                 return genericLogDataRead<OGR_SGFS_NC_Double_Transaction, double>(varId, log);
+#ifdef NETCDF_HAS_NC4
+             case NC_UBYTE:
+                 return genericLogDataRead<OGR_SGFS_NC_UByte_Transaction, unsigned char>(varId, log);
+             case NC_USHORT:
+                 return genericLogDataRead<OGR_SGFS_NC_UShort_Transaction, unsigned short>(varId, log);
+             case NC_UINT:
+                 return genericLogDataRead<OGR_SGFS_NC_UInt_Transaction, unsigned int>(varId, log);
+             case NC_INT64:
+                 return genericLogDataRead<OGR_SGFS_NC_Int64_Transaction, long long>(varId, log);
+             case NC_UINT64:
+                 return genericLogDataRead<OGR_SGFS_NC_UInt64_Transaction, unsigned long long>(varId, log);
+#endif
+             case NC_CHAR:
+             {
+                 size_t readcheck; // 0 means at least one read 0 bytes
+
+                 // Check what type of OP is requested
+                 int8_t op = 0;
+                 readcheck = VSIFReadL(&op, sizeof(int8_t), 1, log);
+                 if(!readcheck)
+                     return MTPtr(); // read failure
+
+                 size_t strsize;
+
+                 // get how much data to read
+                 readcheck = VSIFReadL(&strsize, sizeof(size_t), 1, log);
+                 if(!readcheck)
+                     return MTPtr(); // read failure
+
+                 std::unique_ptr<char, std::default_delete<char[]>> data(new char[strsize + 1]);
```

Comment:
more complicated than needed. Just use `std::string str; str.resize(strsize);`

Reply:
The nice thing about using a char* here is that you have the char* pointer to write to. With an std::string you have a const_char* but you don't have the actual char* representation (which is of course, good encapsulation).

Since we're using VSILReadL / fread to process the binary log, we'll have to store it in some char* array- why not just read them all at once to an actual char* pointer? Otherwise, the thing to do would probably be reading it one by one and packing them into a string object.

Reply:
> you don't have the actual char* representation

not true. If the string is not empty, then &str[0] is the char* representation

Reply:
> 
> 
> > you don't have the actual char* representation
> 
> not true. If the string is not empty, then &str[0] is the char* representation

Okay you got me, but that's kind of breaking the idea of encapsulating a string isn't it?

Reply:
as far as I understand, the C++ standard guarantees that std::string is stored as contiguous char: see https://en.cppreference.com/w/cpp/string/basic_string : `The elements of a basic_string are stored contiguously, that is, for a basic_string s, &*(s.begin() + n) == &*s.begin() + n for any n in [0, s.size()), or, equivalently, a pointer to s[0] can be passed to functions that expect a pointer to the first element of a CharT[] array. `
This is something I heavily rely on.

Reply:
Fine, alright we'll use the string :). Although indexing the pointer of an underlying data structure is still a bit unnerving, you've convinced me that it is at least in the case of the std::string, safe.